core/completion.py

```diff
@@ -619,25 +620,34 @@ def Matches(self, comp):
         # Read the response.  (The name 'COMP_REPLY' would be more consistent with others.)
         val = self.cmd_ev.mem.GetValue('COMPREPLY', scope_e.GlobalOnly)
 
-        if val.tag() == value_e.Undef:
-            # We set it above, so this error would only happen if the user unset it.
-            # Not changing it means there were no completions.
-            # TODO: This writes over the command line; it would be better to use an
-            # error object.
-            print_stderr('osh error: Ran function %r but COMPREPLY was unset' %
-                         self.func.name)
-            return
+        UP_val = val
+        with tagswitch(val) as case:
+            if case(value_e.Undef):
+                # We set it above, so this error would only happen if the user
+                # unset it.  Not changing it means there were no completions.
+                # TODO: This writes over the command line; it would be better
+                # to use an error object.
+                print_stderr('osh error: Ran function %r but COMPREPLY was unset' %
+                             self.func.name)
+                return
 
-        if val.tag() != value_e.BashArray:
-            print_stderr('osh error: COMPREPLY should be an array, got %s' %
-                         ui.ValType(val))
-            return
+            elif case(value_e.Str):
+                val = cast(value.Str, UP_val)
+                strs = [val.s]
+
+            elif case(value_e.BashArray):
+                val = cast(value.BashArray, UP_val)
+                strs = bash_impl.BashArray_GetValues(val)
+
+            else:
+                print_stderr('osh error: COMPREPLY should be an array, got %s' %
```

Comment:
Should be "string or array" now

Reply:
Ah, yes.

Reply:
[`d9b36e6..ab988fb`](https://github.com/oils-for-unix/oils/compare/d9b36e6e45f10ef659680222c195457f3d923547..ab988fb03ad4d4713b22d4cb0964a4f35f366a71)